specs/kyber/src/sampling.rs

```diff
@@ -100,60 +111,60 @@ pub fn sample_from_uniform_distribution(
 ///        = ETA / 2
 /// ```
 ///
-/// This function implements Algorithm 2 of the Kyber Round 3 specification,
-/// which is reproduced below:
+/// This function implements Algorithm 7 of the NIST FIPS 203 standard, which is
+/// reproduced below:
 ///
-/// ```plaintext
-/// Input: Byte array B = (b_0, b_1, . . . , b_{64η−1}) ∈ B^{64η}
-/// Output: Polynomial f ∈ R_q
-/// (β_0, . . . , β_{512η−1}) := BytesToBits(B)
-/// for i from 0 to 255 do
-///     a := sum(j = 0 to η−1)(β_{2iη+j})
-///     b := sum(j = 0 to η−1)(β_{2iη+η+j})
-///     fi := a − b
-/// end for
-/// return f_0 +f_1X +f_2X^2 +···+f_255X^255
-/// ```
+/// Input: byte array B ∈ B^{64η}.
+/// Output: array f ∈ ℤ₂₅₆
 ///
-/// The Kyber Round 3 specification can be found at:
-/// <https://pq-crystals.org/kyber/data/kyber-specification-round3-20210131.pdf>
+/// b ← BytesToBits(B)
+/// for (i ← 0; i < 256; i++)
+///     x ← ∑(j=0 to η - 1) b[2iη + j]
+///     y ← ∑(j=0 to η - 1) b[2iη + η + j]
+///     f[i] ← x−y mod q
+/// end for
+/// return f
 ///
-/// TODO: This requires a different parameter ETA = 3 only when Kyber-512 is
-/// used; generalize it.
-pub fn sample_from_binomial_distribution(
-    sampling_coins: usize,
-    randomness: &[u8],
-) -> KyberPolynomialRingElement {
-    assert_eq!(randomness.len(), sampling_coins * 64);
+/// The NIST FIPS 203 standard can be found at
+/// <https://csrc.nist.gov/pubs/fips/203/ipd>.
+pub fn sample_poly_cbd(eta: usize, bytes: &[u8]) -> KyberPolynomialRingElement {
+    assert_eq!(bytes.len(), eta * 64);
 
-    let random_bits: BitVector = randomness.into();
-    let mut random_bits = random_bits.chunks(sampling_coins);
+    // b ← BytesToBits(B)
+    let bits: BitVector = bytes.into();
 
-    let mut sampled: KyberPolynomialRingElement = KyberPolynomialRingElement::ZERO;
+    let mut bits = bits.chunks(eta);
 
-    for i in 0..sampled.len() {
+    let mut f: KyberPolynomialRingElement = KyberPolynomialRingElement::ZERO;
+
+    for i in 0..f.len() {
         let mut coin_tosses: u8 = 0;
-        for bit in random_bits
+
+        // x ← ∑(j = 0 to η-1) b[2iη + j]
+        for bit in bits
             .next()
             .expect("the assertion ensures there are enough sampling coins")
```

Comment:
I know we looked at these unwrap/expect things before.
But looking at this again I think this makes things a little hard to read.
How about adding a comment and using an unwrap instead? Or even better, hiding the unwrap completely?

Looking at the entire function, we should probably factor out the inner for loops over the `bits` to only have a `sum` function call. Then this looks like the spec.